peps/pep-0747.rst

```diff
@@ -0,0 +1,856 @@
+and can usually be instantiated by calling it.
+
+``TypeExpr`` on the other hand is typically introspected by the user in
+some way, is not necessarily directly instantiable, and is not
+necessarily directly usable in a regular ``isinstance()`` check.
+
+It would be possible to widen ``type`` to include the additional values
+allowed by ``TypeExpr`` but it would reduce clarity about the user’s
+intentions when working with a ``type``. Different concepts and usage
+patterns; different spellings.
+
+.. _designed: https://mail.python.org/archives/list/typing-sig@python.org/message/D5FHORQVPHX3BHUDGF3A3TBZURBXLPHD/
+
+
+Accept arbitrary annotation expressions
+---------------------------------------
+
+Certain typing special forms can be used in *some* but not *all*
+annotation contexts:
+
+For example ``Final[]`` can be used as a variable type but not as a
+parameter type or a return type:
+
+::
+
+   some_const: Final[str] = ...  # OK
+
+   def foo(not_reassignable: Final[object]): ...  # ERROR: Final[] not allowed here
+
+   def nonsense() -> Final[object]: ...  # ERROR: Final[] not meaningful here
+
+``TypeExpr[T]`` does not allow matching such annotation expressions
+because it is not clear how such an expression should constrain a type
+variable in position ``T``:
+
+::
+
+   def ismatch[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]: ...
+
+   def foo(some_arg):
+       if ismatch(some_arg, Final[int]):  # ERROR: Final[int] is not a TypeExpr
+           reveal_type(some_arg)  # ? NOT Final[int], because invalid for a parameter
+
+Functions that wish to operate on *all* kinds of annotation expressions,
+including those that are not ``TypeExpr``\ s, can continue to accept such
+inputs as ``object`` parameters, as they must do so today.
+
+
+Accept only universal type expressions
+--------------------------------------
+
+Earlier drafts of this PEP only allowed ``TypeExpr[]`` to match the subset
+of type expressions which are valid in *all* contexts, excluding
+:ref:`non-universal type expressions <non_universal_typeexpr>`.
+However doing that would effectively
+create a new subset of annotation expressions that Python typing users
+would have to understand, on top of all the existing distinctions between
+“class objects”, “type expressions”, and “annotation expressions”.
+
+To avoid introducing yet another concept that everyone has to learn,
+this proposal just rounds ``TypeExpr[]`` to exactly match the existing
+definition of a “type expression”.
+
+
+Support pattern matching on type expressions
+--------------------------------------------
+
+It was asserted that some functions may wish to pattern match on the
+interior of type expressions in their signatures.
+
+One use case is to allow a function to explicitly enumerate all the
+*specific* kinds of type expressions it supports as input. For example:
+
+::
+
+   @overload
+   def checkcast(typx: TypeExpr[AT=Annotated[T, *Anns]], value: str) -> T: ...
```

Comment:
This is a syntax error, can we just write it without the syntax error?
```suggestion
   def checkcast(typx: TypeExpr[Annotated[T, *Anns]], value: str) -> T: ...
```

Reply:
This is intended to be an example of pattern matching syntax which I'm rejecting.

I'll rephrase `For example:` as `Consider the following possible pattern matching syntax:` to make it more clear that this is not intended to be valid syntax based on what the PEP is proposing.